### PR TITLE
[expo-modules] Create local create-expo-module template

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

We want to easily allow people to create local expo modules.

Instead of:
1. run `create-expo-module`, 2. develop on `example`, 3. publish to npm, 4. npm install, 5. import in your main project

We do:
1. run `create-expo-module --local` 2. import in your main project

# How

Code written by @barthap had already a default config value to autolink modules from `/modules`.
It's not used by anything afaik, and we don't need to change it, so we can just reuse this behaviour.

THIS PR
We expose an easy npx command to create a more minimalistic module template (no example, package.json, eslints, tsconfigs ect) in the correct dir.

https://github.com/expo/expo/pull/21429
We look for index.ts or package.json files in the `/modules/*` directory, when importing `from "some-native-module"`.
Create a local module, run `pod install` or build a dev client and you can import it directly.

https://github.com/expo/expo/pull/21424
Sets up typescript imports for the above.

# Test Plan

Tested manually.

# Checklist

Documenting this needs a new guide/PR.

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/create-expo-module/src/create-expo-module.ts [hunk 1]
```diff
@@ -43,86 +48,116 @@ const IGNORES_PATHS = [
...
-    await fs.remove(path.join(targetDir, 'README.md'));
-  }
-  if (!options.withChangelog) {
-    await fs.remove(path.join(targetDir, 'CHANGELOG.md'));
-  }
-  if (options.example) {
-    // Create "example" folder
-    await createExampleApp(data, targetDir, packageManager);
-  }
+  if (!options.local && data.type !== 'local') {
```
Comment (aleqsio) [inline packages/create-expo-module/src/create-expo-module.ts:124]: data is a tagged union

File: packages/create-expo-module/src/create-expo-module.ts [hunk 2]
```diff
@@ -163,10 +198,10 @@ async function getNpmTarballUrl(packageName: string, version: string = 'latest')
 /**
  * Downloads the template from NPM registry.
  */
-async function downloadPackageAsync(targetDir: string): Promise<string> {
+async function downloadPackageAsync(targetDir: string, isLocal = false): Promise<string> {
   return await newStep('Downloading module template from npm', async (step) => {
     const tarballUrl = await getNpmTarballUrl(
-      'expo-module-template',
+      isLocal ? 'expo-module-template-local' : 'expo-module-template',
```
Comment (aleqsio) [inline packages/create-expo-module/src/create-expo-module.ts:204]: need to publish to NPM

File: packages/expo-module-template-local/ios/{%- project.name %}.podspec [hunk 3]
```diff
@@ -0,0 +1,24 @@
+require 'json'
+
```
Comment (tsapeta) [inline packages/expo-module-template-local/ios/{%- project.name %}.podspec:2]:
```suggestion
```

File: packages/expo-module-template-local/ios/{%- project.name %}.podspec [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+Pod::Spec.new do |s|
+  s.name           = '<%- project.name %>'
+  s.version        = '0.0.1'
+  s.summary        = 'A sample project summary'
+  s.description    = 'A sample project description'
+  s.author         = ''
+  s.homepage       = 'https://expo.dev/'
+  s.platform       = :ios, '13.0'
```
Comment (tsapeta) [inline packages/expo-module-template-local/ios/{%- project.name %}.podspec:10]: Following the podspec syntax reference, it looks like we can remove the platform:
> If not defined in a subspec the attributes of this group inherit the value of the parent.

Generally I would remove every attribute that is not required (better if the user doesn't have to modify the podspec, e.g. when bumping the iOS deployment target in the app)

File: packages/expo-module-template-local/ios/{%- project.name %}.podspec [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  s.dependency 'ExpoModulesCore'
+
+  # Swift/Objective-C compatibility
+  s.pod_target_xcconfig = {
+    'DEFINES_MODULE' => 'YES',
+    'SWIFT_COMPILATION_MODE' => 'wholemodule'
+  }
+  
+  s.source_files = "**/*.{h,m,swift}"
```
Comment (tsapeta) [inline packages/expo-module-template-local/ios/{%- project.name %}.podspec:23]: I guess we can use a wider pattern that includes Objective-C++ and C++. Perhaps we could just include all files, what do you think? 🤔
```suggestion
  s.source_files = "**/*.{h,m,mm,swift,hpp,cpp}"
```

File: packages/create-expo-module/src/utils.ts [hunk 6]
```diff
@@ -1,8 +1,9 @@
 import chalk from 'chalk';
 import spawn from 'cross-spawn';
+import fs from 'fs-extra';
 import githubUsername from 'github-username';
 import ora from 'ora';
-
+import path from 'path';
```
Comment (tsapeta) [inline packages/create-expo-module/src/utils.ts:6]: Move this line up so the imported packages are sorted alphabetically

File: packages/create-expo-module/src/create-expo-module.ts [hunk 7]
```diff
@@ -344,6 +401,19 @@ function printFurtherInstructions(
...
 
+function printFurtherLocalInstructions(slug: string, name: string) {
+  console.log(`You can now import this module inside your application:`);
+  console.log();
+  console.log(chalk.blue(`import { hello } from '${slug}';`));
+  console.log();
+  console.log(`Visit ${chalk.blue.bold(DOCS_URL)} for the documentation on Expo Modules APIs`);
+  console.log(
+    chalk.yellow(
+      `Remember you need to rebuild your development client or reinstall pods to see the changes.`
```
Comment (tsapeta) [inline packages/create-expo-module/src/create-expo-module.ts:412]: Maybe we can ask the user whether to run `pod install` automatically after all?

File: packages/expo-module-template-local/ios/{%- project.name %}.podspec [hunk 8]
```diff
@@ -0,0 +1,24 @@
+require 'json'
+
+Pod::Spec.new do |s|
+  s.name           = '<%- project.name %>'
+  s.version        = '0.0.1'
+  s.summary        = 'A sample project summary'
+  s.description    = 'A sample project description'
+  s.author         = ''
+  s.homepage       = 'https://expo.dev/'
```
[the remark this answers is not in the document]
Comment (aleqsio) [reply packages/expo-module-template-local/ios/{%- project.name %}.podspec:9]: big brain time Tomek!

File: packages/expo-module-template-local/ios/{%- project.name %}.podspec (hunk 5 above)
Comment (aleqsio) [reply packages/expo-module-template-local/ios/{%- project.name %}.podspec:23]: `Perhaps we could just include all files, what do you think? 🤔` - don't have a strong opinion, think I'd keep this as is.

File: packages/create-expo-module/src/create-expo-module.ts (hunk 7 above)
Comment (aleqsio) [reply packages/create-expo-module/src/create-expo-module.ts:412]: Yes, but I'd like to split that to a second PR and consider when we want this to happen.

File: packages/create-expo-module/src/create-expo-module.ts [hunk 9]
```diff
@@ -43,86 +49,116 @@ const IGNORES_PATHS = [
 // Url to the documentation on Expo Modules
 const DOCS_URL = 'https://docs.expo.dev/modules';
 
+const getCorrectLocalDirectory = (targetOrSlug: string) => {
```
Comment (tsapeta) [inline packages/create-expo-module/src/create-expo-module.ts:52]: Make it a function 😅

File: packages/create-expo-module/src/create-expo-module.ts [hunk 10]
```diff
@@ -3,21 +3,27 @@ import chalk from 'chalk';
 import { Command } from 'commander';
 import downloadTarball from 'download-tarball';
 import ejs from 'ejs';
+import { sync } from 'find-up';
```
Comment (tsapeta) [inline packages/create-expo-module/src/create-expo-module.ts:6]: `sync` tells nothing about what the function does.
```suggestion
import findUp from 'find-up';
```

File: packages/create-expo-module/src/create-expo-module.ts [hunk 11]
```diff
@@ -43,86 +49,116 @@ const IGNORES_PATHS = [
 // Url to the documentation on Expo Modules
 const DOCS_URL = 'https://docs.expo.dev/modules';
 
+const getCorrectLocalDirectory = (targetOrSlug: string) => {
+  const packageJsonPath = sync('package.json', { cwd: CWD });
```
Comment (tsapeta) [inline packages/create-expo-module/src/create-expo-module.ts:53]:
```suggestion
  const packageJsonPath = await findUp('package.json', { cwd: CWD });
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**